alibi_detect/saving/_tensorflow/loading.py

```diff
@@ -128,10 +129,9 @@ def load_kernel_config(cfg: dict) -> Callable:
     return kernel
 
 
-def load_optimizer(cfg: dict) -> tf.keras.optimizers.Optimizer:
+def load_optimizer(cfg: dict) -> Union[Type[tf.keras.optimizers.Optimizer], tf.keras.optimizers.Optimizer]:
```

Comment:
`Type[...]` because `optimizer` may be passed to `ClassifierDrift` etc either as an instantiated object or a class. For example, see:

https://github.com/SeldonIO/alibi-detect/blob/6df9bd4b0fb6b27a1df41f0cdae76c804dc3cd55/alibi_detect/cd/tensorflow/classifier.py#L141

As an aside, we should update type hints to show this:
https://github.com/SeldonIO/alibi-detect/blob/6df9bd4b0fb6b27a1df41f0cdae76c804dc3cd55/alibi_detect/cd/tensorflow/classifier.py#L34

Reply:
Serialization logic wrt to above is as follows:

1. If the `optimizer` config is a tensorflow optimizer config, e.g. 
   ```toml
        [optimizer]
        class_name = "Adam"

        [optimizer.config]
        name = "Adam"
        learning_rate = 0.001
        decay = 0.0
   ```
   
   This is deserialized into a `tf.keras.optimizers.Optimizer` object.

2. If the `optimizer` config only specifies the class name e.g. 
      ```toml
        [optimizer]
        class_name = "Adam"
   ```
  Only the class is loaded (i.e. it isn't instantiated). **This approach is the only approach available in the torch implementation**, since torch optimizers cannot be instantiated without providing model parameters.

Reply:
Done in https://github.com/SeldonIO/alibi-detect/pull/656/commits/f3e94d33c140810983a627684622758770b93455.